New Resolver: Second try at implementing constraints

pfmoore opened the pull request:
A reimplementation of constraints for the new resolver. This version extracts the constraints from the root requirement set and handles them independently of "Normal" requirements.

1. When calling `find_matches` on a specifier-based requirement, apply all relevant constraints.
2. When calling `find_matches` on an explicit requirement, fail if any constraints apply. This is needed to implement the existing behaviour of failing with an error if a constraint applies to a path or URL based requirement.

At the moment, the constraints are just dumped on the factory object. This is simply to prove that the approach works. For the final implementation, we need to find somewhere more appropriate to store the constraints. The provide seems the best option, but then we need to pass the constraints through to `find_matches`, and I'm not sure we want to require *every* `Requirement` subclass to take a `constraints` argument in `find_matches...

pfmoore commented:
@uranusjr @pradyunsg I'd appreciate comments, particularly on the approach and on anything I might have missed...

uranusjr commented:
The approach seems straightforward to me. I’m wondering if this would affect error reporting, because contraints don’t show up in the resolver’s recorded requirements.

pfmoore commented:
> I’m wondering if this would affect error reporting

Possibly. The best way to find out would be to write (or identify) some tests that exercise error reporting with constraints involved. (That's how I found the "installation from path or url cannot be constrained to a version" issue). Unfortunately, I don't really know any good way to find such cases. Maybe it's something @ilanschnell could consider as part of his testing work?

pradyunsg commented on tests/functional/test_new_resolver.py:
```diff
@@ -529,3 +529,54 @@ def test_new_resolver_handles_prerelease(
+
+def test_new_resolver_constraint_on_dependency(script):
+    create_basic_wheel_for_package(script, "base", "1.0", depends=["dep"])
+    create_basic_wheel_for_package(script, "dep", "1.0")
```
```suggestion
    create_basic_wheel_for_package(script, "dep", "0.1")
    create_basic_wheel_for_package(script, "dep", "1.0")
    create_basic_wheel_for_package(script, "dep", "2.0")
```

To actually test that the resolver didn't just pick the first compatible version, but rather the specific pin.

pfmoore replied:
Good catch, I like that! (I'd hope that there's already a "real" test that covers this, but I don't mind a bit of redundancy as we don't yet have the full test suite sufficiently under control for that to be a realistic option).

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

pfmoore commented:
OK, marking this as ready to go. As discussed, I'll work on moving the constraints (and root requirements) off the factory object and onto the provider in a follow-up PR.

uranusjr commented:
Damn, I just realised I reviewed the wrong PR and wrote all my comments on the refactoring one 😞 Most of the comments should be here instead. Maybe I’ll copy them here and delete those from #8195.

uranusjr commented on src/pip/_internal/resolution/resolvelib/requirements.py:
```diff
@@ -35,6 +37,12 @@ def name(self):
+                "Could not satisfy constraints for '{}': "
+                "installation from path or url cannot be "
+                "constrained to a version".format(self.name)
+            )
```
Is this to match the legacy resolver’s behaviour? Because to me it makes sense to allow this (if the version satisfies the constraints, of course).

pfmoore replied:
Yes, it's to match the legacy resolver.

I don't mind changing this, but I'd rather not do it in either of these two PRs. As a behaviour change it deserves its own PR (and probably even a news item!)

uranusjr replied:
Makes sense to me, this would be something to tell the world about 📰

uranusjr commented on src/pip/_internal/resolution/resolvelib/resolver.py:
```diff
@@ -69,9 +71,21 @@ def resolve(self, root_reqs, check_supported_wheels):
+                assert req.name
+                assert req.specifier
+                name = canonicalize_name(req.name)
+                constraints[name].append(req.specifier)
```
I wonder if it is possible to merge the specifiers here. It would make the contraint checking code more readable since the loops can be eliminated.

pfmoore replied:
Hmm, I can look at that. Good suggestion.

pfmoore replied:
Yes we can. `SpecifierSet & SpecifierSet` does what we want. I'll fix this tomorrow. It's getting late and I don't want to mess things up, with the refactor PR being based on this one.

uranusjr commented on src/pip/_internal/resolution/resolvelib/resolver.py:
```diff
@@ -69,9 +71,21 @@ def resolve(self, root_reqs, check_supported_wheels):
-        # FIXME: Implement constraints.
-        if any(r.constraint for r in root_reqs):
-            raise InstallationError("Constraints are not yet supported.")
+        constraints = defaultdict(list)  # type: Dict[str,List[SpecifierSet]]
```
This uses `defaultdict` but the provider uses `dict.get()` to access. Only one of them is required.

pfmoore replied:
Never though of that! I typed it as `Dict` though - does that matter? I'm using `defaultdict` because it's easier for construction, but once constructed I want to treat it as a dictionary (although I'm fine with omitting the `get` call).

uranusjr replied:
Hmm, yeah I guess that makes sense if you type it as `Dict`. I *think* `defaultdict.__getitem__()` is more performance than `dict.get()`, but that’s too insignificant to matter anyway. Let’s keep this as-is.

pfmoore replied:
Hmm. On reflection, I'd rather keep the `.get()`, just in case someone later refactors the code that builds the constraints and uses a normal dictionary rather than a defaultdict.

pfmoore commented:
I've pulled the change to move constraints off the factory from the other PR into this one, because it made it easier to avoid merge conflicts.

I've checked and I think I've addressed all the comments now, so if you can take another look and let me know if this looks good to go.

uranusjr commented on src/pip/_internal/resolution/resolvelib/requirements.py:
```diff
@@ -33,8 +34,14 @@ def name(self):
+                "Could not satisfy constraints for '{}': "
+                "installation from path or url cannot be "
+                "constrained to a version".format(self.name)
+            )
```
Is this stale? The provider is now passing in `SpecifierSet`, not `Sequence[SpecifierSet]`. Also the `len()` part works? 🤔

pfmoore replied:
`SpecifierSet` implements `__len__` to give the number of specifiers in the set. See [here](https://packaging.pypa.io/en/latest/specifiers/#packaging.specifiers.SpecifierSet.__len__).

I was quite pleased when I found this 🙂

uranusjr commented on src/pip/_internal/resolution/resolvelib/requirements.py:
```diff
@@ -104,8 +116,10 @@ def name(self):
+    def find_matches(self, constraint):
+        # type: (Sequence[SpecifierSet]) -> Sequence[Candidate]
+        assert len(constraint) == 0, \
+            "RequiresPythonRequirement cannot have constraints"
```
Same here, the argument type seems wrong.

pfmoore replied:
The type hint does look wrong here. I may have messed up a merge. Not sure how mypy let me get away with that... I'll fix it.

pradyunsg replied:
mypy is only as good as the annotations. And, honestly, not very good at incremental typing. :(

pfmoore replied:
Yeah, no worries. I just prefer to blame my tools than admit I made a mistake :-)

pfmoore commented:
@prometheanfire @dhellmann The new constraint implementation has landed in master. If you can test it out on your environment, that would be awesome 🙂 

You'll need to pull pip from master - `py -m pip install "git+https://github.com/pypa/pip"` should do the job. And you'll need to either add the `--unstable-feature=resolver` flag when running pip, or set `$env:PIP_UNSTABLE_FEATURE="resolver"`

Shout out if you have any difficulties!

prometheanfire commented:
testing here, we'll see if it works

https://review.opendev.org/726186
